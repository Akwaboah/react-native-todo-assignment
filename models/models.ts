export interface ITodoItem {
  'id':number,
  'task': String,
}
export enum ApiStatus {
  LOADING = 'loading',
  LOADED = 'loaded',
  ADDING = 'Adding',
  DELETE = 'delete',
  FAILED = 'failed',
}
