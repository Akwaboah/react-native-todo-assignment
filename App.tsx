/* eslint-disable react-native/no-inline-styles */
import 'reflect-metadata';
import React from 'react';
import {
  KeyboardAvoidingView,
  StyleSheet,
  Text,
  View,
  TextInput,
  TouchableOpacity,
  Keyboard,
  ScrollView,
  Platform,
  ToastAndroid,
} from 'react-native';
import { injectable } from 'tsyringe';
import Task from './components/Task';
import {ApiStatus, ITodoItem} from './models/models';

@injectable()
class App extends React.Component<AppProps, {task: String}> {
  constructor(props: AppProps | Readonly<AppProps>) {
    super(props);
    this.state = {task: ''};
  }

  componentDidMount() {
    this.props.loadTodos();
  }

  handleAddTask = () => {
    let taskData = this.state.task;
    if (taskData.trim() === '') {
      ToastAndroid.showWithGravity('Please enter your task',ToastAndroid.SHORT,
        ToastAndroid.CENTER,
      );
      return;
    }
    this.props.addTodo(taskData);
    this.setState({task: ''});
    Keyboard.dismiss();
  };

  OnchangeTask = (val: String) => {
    this.setState({task: val});
  };

  RemoveTask = (todo_id: number) => {
    this.props.deleteTodo(todo_id);
  };

  render() {
    const {todos, loadingStatus} = this.props;
    return (
      <View style={styles.container}>
        <ScrollView
          contentContainerStyle={{flexGrow: 1}}
          keyboardShouldPersistTaps="handled">
          <View style={styles.tasksWrapper}>
            <Text style={styles.sectionTitle}>My Daily Task</Text>
            <View style={styles.items}>
              {loadingStatus === ApiStatus.LOADING && (
                <View
                  style={{
                    flex: 1,
                    justifyContent: 'center',
                    alignItems: 'center',
                  }}>
                  <Text
                    style={{color: 'green', fontFamily: 'bold', fontSize: 18}}>
                    Please wait, fetching task...
                  </Text>
                </View>
              )}
              {loadingStatus === ApiStatus.FAILED && (
                <View
                  style={{
                    flex: 1,
                    justifyContent: 'center',
                    alignItems: 'center',
                  }}>
                  <Text
                    style={{color: 'red', fontFamily: 'bold', fontSize: 18}}>
                    Connection error...
                  </Text>
                </View>
              )}
              {loadingStatus === ApiStatus.LOADED && todos.length === 0 && (
                <View
                  style={{
                    flex: 1,
                    justifyContent: 'center',
                    alignItems: 'center',
                  }}>
                  <Text style={{fontFamily: 'bold', fontSize: 18}}>
                    Task empty, add new ....
                  </Text>
                </View>
              )}

              {loadingStatus === ApiStatus.LOADED &&
                todos.length > 0 &&
                todos.map(todo => (
                  <TouchableOpacity
                    key={todo.id}
                    onPress={() => {
                      this.RemoveTask(todo.id);
                    }}>
                    <Task text={todo.task} />
                  </TouchableOpacity>
                ))}
            </View>
          </View>
        </ScrollView>
        <KeyboardAvoidingView
          behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
          style={styles.writeTaskWrapper}>
          <TextInput
            style={styles.input}
            value={this.state.task.toString()}
            placeholder={'enter task here'}
            onChangeText={text => this.OnchangeTask(text)}
          />
          <TouchableOpacity testID="add_task" onPress={() => this.handleAddTask()}>
            <View style={styles.addWrapper}>
              <Text style={styles.addText}>+</Text>
            </View>
          </TouchableOpacity>
        </KeyboardAvoidingView>
      </View>
    );
  }
}

export default App;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#E8EAED',
  },
  tasksWrapper: {
    paddingTop: 10,
    paddingHorizontal: 20,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: 'bold',
  },
  items: {
    marginTop: 8,
  },
  writeTaskWrapper: {
    position: 'absolute',
    bottom: 20,
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  input: {
    paddingVertical: 15,
    paddingHorizontal: 15,
    backgroundColor: '#FFF',
    borderRadius: 5,
    borderColor: '#C0C0C0',
    borderWidth: 1,
    width: 250,
    height: 45,
  },
  addWrapper: {
    width: 60,
    height: 45,
    backgroundColor: '#FFF',
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    borderColor: '#C0C0C0',
    borderWidth: 1,
  },
  addText: {},
});

// Define props coming from redux store
export interface IAppStateProps {
  loadingStatus: ApiStatus;
  todos: ITodoItem[];
}

// Define props that are action creators
export interface IAppDispatchProps {
  loadTodos: () => void;
  addTodo: (task_description: String) => void;
  deleteTodo: (id: number) => void;
}

type AppProps = IAppStateProps & IAppDispatchProps;
