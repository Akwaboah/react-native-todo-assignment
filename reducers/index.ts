import { combineReducers } from 'redux';
import todosReducer, {ITodoState, initialTodoState} from './todoReducer';

export interface IState {
  todos: ITodoState;
}

export const initialState = (
  {todos: initialTodoState}
);

const rootReducer = combineReducers({
  todos: todosReducer,
});

export default rootReducer;
