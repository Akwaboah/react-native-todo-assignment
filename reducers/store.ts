import { createEpicMiddleware } from 'redux-observable';
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import rootEpic from '../epics/index';

import rootReducer, { initialState } from './index';

const composeEnhancer = composeWithDevTools({
  name: 'React Clean Architecture',
});

const epicMiddleware = createEpicMiddleware();

const store = createStore(
  rootReducer,
  initialState,
  composeEnhancer(applyMiddleware(epicMiddleware))
);

epicMiddleware.run(rootEpic);

export default store;
