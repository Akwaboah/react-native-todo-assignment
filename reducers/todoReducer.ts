import produce from 'immer';
import { TodosAction, TodosActionTypes } from '../actions/todoActions';
import { ApiStatus, ITodoItem } from '../models/models';

export interface ITodoState {
    loadingStatus: ApiStatus;
    addingStatus: ApiStatus;
    deletingStatus: ApiStatus;
    todos: ITodoItem[];
}

export const initialTodoState: ITodoState = {
    loadingStatus: ApiStatus.LOADING,
    addingStatus: ApiStatus.LOADED,
    deletingStatus: ApiStatus.DELETE,
    todos: [],
};

export default function todosReducer(state: ITodoState = initialTodoState, action: TodosAction) {
    return produce(state, draft => {
      switch (action.type) {
        case TodosActionTypes.LOAD_TODOS:
        case TodosActionTypes.LOADING_TODOS:
          draft.loadingStatus = ApiStatus.LOADING;
          break;

        case TodosActionTypes.LOADING_TODOS_FAILED:
          draft.loadingStatus = ApiStatus.FAILED;
          break;

        case TodosActionTypes.LOADED_TODOS:
          draft.loadingStatus = ApiStatus.LOADED;
          draft.todos = action.payload.todos;
          break;

        case TodosActionTypes.ADD_TODO:
        case TodosActionTypes.ADDING_TODO:
          draft.addingStatus = ApiStatus.LOADING;
          break;

        case TodosActionTypes.ADDING_TODOS_FAILED:
          draft.addingStatus = ApiStatus.FAILED;
          break;

        case TodosActionTypes.ADDED_TODOS:
          draft.todos.push(action.payload.todos);
          break;

        case TodosActionTypes.DELETE_TODO:
          draft.loadingStatus = ApiStatus.DELETE;
          draft.todos.filter(todo => todo.id !== action.payload.index);
          break;
      }
    });
}
