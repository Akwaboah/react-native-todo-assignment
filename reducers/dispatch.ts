import { IAppStateProps, IAppDispatchProps } from '../App';
import { IState } from '../reducers/index';
import { addTodo, deleteTodo, loadTodos } from '../actions/todoActions';

//am using this file to handle props and state attachment to my redux component

export function mapStateToDispatch(state: IState): IAppStateProps {
    return {
        loadingStatus: state.todos.loadingStatus,
        todos: state.todos.todos,
    };
}

export const mapDispatchToProps: IAppDispatchProps = {
    addTodo,
    deleteTodo,
    loadTodos,
};
