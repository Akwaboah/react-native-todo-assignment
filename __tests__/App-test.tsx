import React from 'react';
import {Provider} from 'react-redux';
import store from '../reducers/store';
import {connect} from 'react-redux';
import App from '../App';
import {mapStateToDispatch, mapDispatchToProps} from '../reducers/dispatch';
import {act, create, ReactTestRenderer} from 'react-test-renderer';
import {IState} from '../reducers';
import {ApiStatus} from '../models/models';

const ConnectedComponent = connect(mapStateToDispatch, mapDispatchToProps)(App);

describe('Main app to match first snapshot', () => {

  it('Must match snapshoot', done => {
    let wrapper: ReactTestRenderer;
    act(() => {
      wrapper = create(
        <Provider store={store}>
          <ConnectedComponent />
        </Provider>,
      );
    });
    setTimeout(() => {
      // expect(wrapper.getByText('My Todos')).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
      done();
    }, 1000);
  });

  let mockState: IState;

  beforeEach(() => {
    // mockDispatch = {loadTodos:jest.fn(),addTodo:jest.fn(), deleteTodo:jest.fn()};
    mockState = {
      todos: {
        loadingStatus: ApiStatus.LOADING,
        addingStatus: ApiStatus.ADDING,
        deletingStatus: ApiStatus.DELETE,
        todos: [
          {
            id: 2,
            task: 'String',
          },
        ],
      },
    };
  });

  it('maps state to props correctly', () => {
    const props = mapStateToDispatch(mockState);
    expect(props.loadingStatus).toEqual(mockState.todos.loadingStatus);
  });
});
