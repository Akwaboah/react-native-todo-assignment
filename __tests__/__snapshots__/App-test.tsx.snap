// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main app to match first snapshot Must match snapshoot 1`] = `
<View
  style={
    {
      "backgroundColor": "#E8EAED",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
  >
    <View>
      <View
        style={
          {
            "paddingHorizontal": 20,
            "paddingTop": 10,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          My Daily Task
        </Text>
        <View
          style={
            {
              "marginTop": 8,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                {
                  "color": "green",
                  "fontFamily": "bold",
                  "fontSize": 18,
                }
              }
            >
              Please wait, fetching task...
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "bottom": 20,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "position": "absolute",
          "width": "100%",
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="enter task here"
      style={
        {
          "backgroundColor": "#FFF",
          "borderColor": "#C0C0C0",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 45,
          "paddingHorizontal": 15,
          "paddingVertical": 15,
          "width": 250,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="add_task"
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFF",
            "borderColor": "#C0C0C0",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 45,
            "justifyContent": "center",
            "width": 60,
          }
        }
      >
        <Text
          style={{}}
        >
          +
        </Text>
      </View>
    </View>
  </View>
</View>
`;
